File: Ant/chapter3/python/work_assignment1.py
INF = int(1e9)


def ns(f):
    return next(f).strip()


class Edge:
    def __init__(self, v, cap, rev):
        self.v = v
        self.cap = cap
        self.rev = rev


with open("../testset/work_assignment/test2.txt", 'r') as f:
    N, M, K = map(int, ns(f).split())
    G = [[] for _ in range(N + M + 2)]

    def add_edge(u, v, cap):
        global G
        G[u].append(Edge(v, cap, len(G[v])))
        G[v].append(Edge(u, 0, len(G[u]) - 1))

    for _ in range(K):
        u, v = map(int, ns(f).split())
        add_edge(u - 1, N + v - 1, 1)


def dfs(u, t, f, used):
    global G
    if u == t:
        return f
    used[u] = True
    for e in G[u]:
        if not used[e.v] and e.cap > 0:
            d = dfs(e.v, t, min(f, e.cap), used)
            if d > 0:
                e.cap -= d
                G[e.v][e.rev].cap += d
                return d
    return 0


s = N + M
t = s + 1
for i in range(N):
    add_edge(s, i, 1)
for i in range(M):
    add_edge(N + i, t, 1)
res = 0
while True:
    used = [False] * (N + M + 2)
    f = dfs(s, t, INF, used)
    if f == 0:
        print(res)
        exit()
    res += f

File: Ant/chapter3/cpp/work_assignment1.cpp
#include <algorithm>
#include <fstream>
#include <memory.h>
#include <stdio.h>
#include <vector>

#define MAX_N 1000
#define MAX_M 1000
#define INF 1'000'000'000

using namespace std;

struct Edge {
  int v, cap, rev;
};

int N, M, K;
vector<Edge> G[MAX_N + MAX_M + 2]; // 0..N-1:=computer, N..N+K-1:=work, N+K:=s, N+K+1:=t
bool used[MAX_N + MAX_M + 2];

void add_edge(int u, int v, int cap) {
  G[u].push_back(Edge{v, cap, G[v].size()});
  G[v].push_back(Edge{u, 0, G[u].size() - 1});
}

int dfs(int u, int t, int f) {
  if (u == t) {
    return f;
  }
  used[u] = true;
  for (int i = 0; i < G[u].size(); ++i) {
    Edge &e = G[u][i];
    if (!used[e.v] && e.cap > 0) {
      int d = dfs(e.v, t, min(f, e.cap));
      if (d > 0) {
        e.cap -= d;
        G[e.v][e.rev].cap += d;
        return d;
      }
    }
  }
  return 0;
}

int main() {
  ifstream ifs("../testset/work_assignment/test2.txt");
  ifs >> N >> M >> K;
  for (int i = 0; i < K; ++i) {
    int u, v;
    ifs >> u >> v;
    add_edge(u - 1, N + v - 1, 1);
  }
  int s = N + M;
  int t = s + 1;
  for (int i = 0; i < N; ++i) {
    add_edge(s, i, 1);
  }
  for (int i = 0; i < M; ++i) {
    add_edge(N + i, t, 1);
  }
  int res = 0;
  for (;;) {
    memset(used, 0, sizeof(used));
    int f = dfs(s, t, INF);
    if (f == 0) {
      printf("%d\n", res);
      return 0;
    }
    res += f;
  }
}